### PR TITLE
Consolidate auth APIs

### PR DESCRIPTION
**Pull request by Tratcher**
#266 - Consolidate all the auth APIs to HttpContext.Authentication
#270 - Naming
@davidfowl @muratg 
/cc @HaoK

### CODE REVIEW

**Comment by Tratcher** [comment]
Update to include #273 - POCO context objects.

File: src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                new AuthenticationProperties(authenticateContext.Properties),
+                new AuthenticationDescription(authenticateContext.Description));
+        }
+
+        public override void Challenge(AuthenticationProperties properties, string authenticationScheme)
+        {
+            HttpResponseFeature.StatusCode = 401;
+            var handler = HttpAuthenticationFeature.Handler;
+
+            var challengeContext = new ChallengeContext(authenticationScheme, properties == null ? null : properties.Items);
```
Comment (HaoK) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: properties?.Items

File: src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            }
+
+            // Verify all types ack'd
+            if (!signInContext.Accepted)
+            {
+                throw new InvalidOperationException("The following authentication scheme was not accepted: " + authenticationScheme);
+            }
+        }
+
+        public override void SignOut(string authenticationScheme, AuthenticationProperties properties)
```
Comment (HaoK) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: Should we introduce a new POCO that contains both the scheme and AuthenticationProperties, they always seem to be used together...  Or maybe just push the Scheme inside of the properties as a special property that's required?

```
public class AuthenticationProperties {
   AuthenticationProperties(string scheme)

   string Scheme { get; }
   Dictionary<string, string> Items { get; }
}

Comment (Tratcher) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: I've seen enough examples of scheme being used without properties that I think it's more convenient to keep them separate.

File: src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            var authenticateContext = new AuthenticateContext(authenticationScheme);
+            if (handler != null)
+            {
+                await handler.AuthenticateAsync(authenticateContext);
+            }
+
+            // Verify all types ack'd
+            if (!authenticateContext.Accepted)
+            {
+                throw new InvalidOperationException("The following authentication scheme was not accepted: " + authenticationScheme);
```
Comment (muratg) [inline src/Microsoft.AspNet.Http/Authentication/DefaultAuthenticationManager.cs]: Nit: $"The following.... {authenticationScheme}.."

Ditto other exception messages below.

**Comment by Tratcher** [comment]
Updated and rebased.

**Comment by Tratcher** [comment]
Updated with related changes for WebSockets. #267

File: src/Microsoft.AspNet.Http/DefaultWebSocketManager.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        private IHttpWebSocketFeature WebSocketFeature
+        {
+            get { return _webSockets.Fetch(_features); }
+        }
+
+        public override bool IsWebSocketRequest
+        {
+            get
+            {
+                var webSocketFeature = WebSocketFeature;
```
Comment (muratg) [inline src/Microsoft.AspNet.Http/DefaultWebSocketManager.cs]: Why not use the property directly? Do we need to assign it to a variable?

File: src/Microsoft.AspNet.Http/DefaultWebSocketManager.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            get
+            {
+                return ParsingHelpers.GetHeaderUnmodified(HttpRequestFeature.Headers,
+                    HeaderNames.WebSocketSubProtocols) ?? EmptyList;
+            }
+        }
+
+        public override Task<WebSocket> AcceptWebSocketAsync(string subProtocol)
+        {
+            var webSocketFeature = WebSocketFeature;
```
Comment (muratg) [inline src/Microsoft.AspNet.Http/DefaultWebSocketManager.cs]: Looks like this is not used.

**Comment by muratg** [comment]
Code changes look good to me and reflects the discussions from the API review.  

@davidfowl does new API surface look good to you?

File: src/Microsoft.AspNet.Http.Core/Authentication/AuthenticationManager.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            Challenge(properties, "");
+        }
+
+        public virtual void Challenge(string authenticationScheme)
+        {
+            Challenge(properties: null, authenticationScheme: authenticationScheme);
+        }
+
+        public abstract void Challenge(AuthenticationProperties properties, string authenticationScheme);
```
Comment (HaoK) [inline src/Microsoft.AspNet.Http.Core/Authentication/AuthenticationManager.cs]: I agree with Brent's issue, we should probably switch scheme to be the first parameter here to be consistent with sign in/signout

**Pull request merged**